schema/kotlin-core/src/main/kotlin/br/com/zup/beagle/builder/BeagleBuilder.kt

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package br.com.zup.beagle.builder
+
+@BeagleDsl
+interface BeagleBuilder<T> {
+
+    fun <F> listBuilder(block: BeagleListBuilder<F>.() -> Unit)
+            = BeagleListBuilder<F>().apply(block).build()
+
+    fun <F> listBuilderNullable(block: BeagleListBuilder<F>.() -> Unit)
+            = BeagleListBuilder<F>().apply(block).buildNullable()
+
+    fun <K : Any, F> mapBuilder(block: BeagleMapBuilder<K, F>.() -> Unit)
+            = BeagleMapBuilder<K, F>().apply(block).build()
+
+    fun <K : Any, F> mapBuilderNullable(block: BeagleMapBuilder<K, F>.() -> Unit)
+            = BeagleMapBuilder<K, F>().apply(block).buildNullable()
+
+    fun build(): T
+}
+
+class BeagleListBuilder<T> {
+    private var list: MutableList<T>? = null
+
+    fun list(list: List<T>?) {
+        if (list != null) addList(list)
+    }
+
+    fun entry(value: T?) {
+        if (value != null) addEntry(value)
+    }
+
+    fun list(block: () -> List<T>?) {
+        list(block.invoke())
+    }
+
+    fun entry(block: () -> T?) {
+        entry(block.invoke())
+    }
+
+    operator fun T.unaryPlus() {
+        entry(this)
+    }
+
+    operator fun List<T>.unaryPlus() {
+        list(this)
+    }
+
+    fun build() = list!!.toList()
```

Comment:
Is it really necessary to throw an NPE? If not, how about using the Elvis operator?
```suggestion
    fun build() = list?.toList() ?: emptyList()
```

Reply:
It's definitely not necessary and I can change that, but I feel like this should be an user's logic as he is assigning a value to a non nullable field, no? What do you think?

Reply:
In this case you could throw a more explanatory exception
```suggestion
    fun build() = list?.toList() ?: throw Exception("do this instead")
```

Reply:
I'll do that!